Emby.Server.Implementations/Chapters/ChapterManager.cs

```diff
@@ -0,0 +1,316 @@
+            return false;
+        }
+
+        if (!video.IsCompleteMedia)
+        {
+            return false;
+        }
+
+        // Can't extract images if there are no video streams
+        return video.DefaultVideoStreamIndex.HasValue;
+    }
+
+    private long GetAverageDurationBetweenChapters(IReadOnlyList<ChapterInfo> chapters)
+    {
+        if (chapters.Count < 2)
+        {
+            return 0;
+        }
+
+        long sum = 0;
+        for (int i = 1; i < chapters.Count; i++)
+        {
+            sum += chapters[i].StartPositionTicks - chapters[i - 1].StartPositionTicks;
+        }
+
+        return sum / chapters.Count;
+    }
+
+    /// <inheritdoc />
+    public async Task<bool> RefreshChapterImages(Video video, IDirectoryService directoryService, IReadOnlyList<ChapterInfo> chapters, bool extractImages, bool saveChapters, CancellationToken cancellationToken)
+    {
+        if (chapters.Count == 0)
+        {
+            return true;
+        }
+
+        var libraryOptions = _libraryManager.GetLibraryOptions(video);
+
+        if (!IsEligibleForChapterImageExtraction(video, libraryOptions))
+        {
+            extractImages = false;
+        }
+
+        var averageChapterDuration = GetAverageDurationBetweenChapters(chapters);
+        var threshold = TimeSpan.FromSeconds(1).Ticks;
+        if (averageChapterDuration < threshold)
+        {
+            _logger.LogInformation("Skipping chapter image extraction for {Video} as the average chapter duration {AverageDuration} was lower than the minimum threshold {Threshold}", video.Name, averageChapterDuration, threshold);
+            extractImages = false;
+        }
+
+        var success = true;
+        var changesMade = false;
+
+        var runtimeTicks = video.RunTimeTicks ?? 0;
+
+        var currentImages = GetSavedChapterImages(video, directoryService);
+
+        foreach (var chapter in chapters)
+        {
+            if (chapter.StartPositionTicks >= runtimeTicks)
+            {
+                _logger.LogInformation("Stopping chapter extraction for {0} because a chapter was found with a position greater than the runtime.", video.Name);
+                break;
+            }
+
+            var path = _pathManager.GetChapterImagePath(video, chapter.StartPositionTicks);
+
+            if (!currentImages.Contains(path, StringComparison.OrdinalIgnoreCase))
+            {
+                if (extractImages)
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+
+                    try
+                    {
+                        // Add some time for the first chapter to make sure we don't end up with a black image
+                        var time = chapter.StartPositionTicks == 0 ? TimeSpan.FromTicks(Math.Min(_firstChapterTicks, video.RunTimeTicks ?? 0)) : TimeSpan.FromTicks(chapter.StartPositionTicks);
+
+                        var inputPath = video.Path;
+                        if (!string.IsNullOrEmpty(path))
```

Comment:
This can't be `null` or empty, can it?

Reply:
This is a straight up copy from the old code. I think path is nullable in general though?

Reply:
I don't see it
https://github.com/jellyfin/jellyfin/pull/13847/files#diff-0765f49fee34b027c599638ac97dfa364e897edce99deda73a99631a171bdf19L162